File: web-admin/src/api/inventory.ts
import { api } from './http';

export interface InventoryCategory {
    id: string;
    code: string;
    name: string;
    description: string | null;
    inventory_account_id: string | null;
    expense_account_id: string | null;
    item_count?: number; // Optional count from backend if available
    created_at: string;
    updated_at: string;
}

export interface InventoryItem {
    id: string;
    category_id: string;
    unit_id: number;
    sku: string;
    name: string;
    description: string | null;
    min_stock: number;
    max_stock: number;
    initial_quantity: number;
    current_quantity: number;
    purchase_price: number;
    average_cost: number;
    is_active: boolean;
    created_at: string;
    updated_at: string;
    category_name?: string;
}

export interface CreateInventoryCategoryRequest {
    code: string;
    name: string;
    description?: string;
    inventory_account_id?: string;
    expense_account_id?: string;
}

export interface UpdateInventoryCategoryRequest {
    name?: string;
    description?: string;
    inventory_account_id?: string;
    expense_account_id?: string;
}

export const inventoryApi = {
    // Categories
    listCategories: async () => {
        const response = await api.get('/inventory/categories');
        return response.data.data;
    },
    getCategory: async (id: string) => {
        const response = await api.get(`/inventory/categories/${id}`);
        return response.data.data;
    },
    createCategory: async (data: CreateInventoryCategoryRequest) => {
        const response = await api.post('/inventory/categories', data);
        return response.data.data;
    },
    updateCategory: async (id: string, data: UpdateInventoryCategoryRequest) => {
        const response = await api.put(`/inventory/categories/${id}`, data);
        return response.data.data;
    },
    deleteCategory: async (id: string) => {
        const response = await api.delete(`/inventory/categories/${id}`);
        return response.data.data;
    },

    // Items
    listItems: async (params?: any) => {
        const response = await api.get('/inventory/items', { params });
        return response.data.data;
    },
    getItem: async (id: string) => {
        const response = await api.get(`/inventory/items/${id}`);
        return response.data.data;
    },
    createItem: async (data: any) => {
        const response = await api.post('/inventory/items', data);
        return response.data.data;
    },
    updateItem: async (id: string, data: any) => {
        const response = await api.put(`/inventory/items/${id}`, data);
        return response.data.data;
    },
    deleteItem: async (id: string) => {
        const response = await api.delete(`/inventory/items/${id}`);
        return response.data.data;
    },
    listMovements: async (params?: { item_id?: string; limit?: number }) => {
        const response = await api.get('/inventory/movements', { params });
        return response.data.data;
    },
    // Add other item methods as needed
    batchAdjust: async (data: BatchInventoryAdjustmentRequest) => {
        const response = await api.post('/inventory/adjust/batch', data);
        return response.data.data;
    },

    bulkCreate: async (data: { items: any[] }) => {
        const response = await api.post('/inventory/items/bulk', data);
        return response.data.data;
    },

    // Documents
    listDocuments: async (id: string) => {
        const response = await api.get(`/inventory/items/${id}/documents`);
        return response.data.data;
    },
    addDocument: async (id: string, data: any) => {
        const response = await api.post(`/inventory/items/${id}/documents`, data);
        return response.data.data;
    },
    uploadFile: async (file: File) => {
        const formData = new FormData();
        formData.append('file', file);
        const response = await api.post('/upload', formData, {
            headers: { 'Content-Type': 'multipart/form-data' },
        });
        return response.data;
    },
};

export interface InventoryDocument {
    id: string;
    item_id: string;
    name: string;
    type: string;
    file_path: string;
    mime_type: string | null;
    size_bytes: number | null;
    expiry_date: string | null;
    notes: string | null;
    uploaded_by: string | null;
    created_at: string;
    updated_at: string;
}

export interface InventoryMovement {
    id: string;
    item_id: string;
    movement_type: 'IN_PURCHASE' | 'IN_ADJUSTMENT' | 'OUT_USAGE' | 'OUT_ADJUSTMENT' | 'OUT_TRANSFER';
    quantity: number;
    unit_price: number;
    total_value: number;
    reference_id?: string;
    reference_number?: string;
    notes?: string;
    created_by?: string;
    created_at: string;
}

export interface BatchInventoryAdjustmentRequest {
    items: InventoryAdjustmentItem[];
    notes?: string;
}

export interface InventoryAdjustmentItem {
    item_id: string;
    quantity: number;
    unit_price?: number;
}

File: src/domain/entities/inventory.rs
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct InventoryCategory {
    pub id: Uuid,
    pub code: String,
    pub name: String,
    pub description: Option<String>,
    pub inventory_account_id: Option<Uuid>,
    pub expense_account_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct InventoryItem {
    pub id: Uuid,
    pub category_id: Uuid,
    pub unit_id: i32,
    pub sku: String,
    pub name: String,
    pub description: Option<String>,
    pub min_stock: Decimal,
    pub max_stock: Decimal,
    pub current_quantity: Decimal,
    pub average_cost: Decimal,
    pub last_purchase_price: Decimal,
    pub is_active: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "inventory_movement_type")]
pub enum InventoryMovementType {
    #[serde(rename = "IN_PURCHASE")]
    InPurchase,
    #[serde(rename = "IN_ADJUSTMENT")]
    InAdjustment,
    #[serde(rename = "OUT_USAGE")]
    OutUsage,
    #[serde(rename = "OUT_ADJUSTMENT")]
    OutAdjustment,
    #[serde(rename = "OUT_TRANSFER")]
    OutTransfer,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct InventoryMovement {
    pub id: Uuid,
    pub item_id: Uuid,
    pub movement_type: InventoryMovementType,
    pub quantity: Decimal,
    pub unit_price: Decimal,
    pub total_value: Decimal,
    pub reference_id: Option<Uuid>,
    pub reference_number: Option<String>,
    pub notes: Option<String>,
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}
